File: intern/1750.minimum-length-of-string-after-deleting-similar-ends.js
/*
 * @lc app=leetcode id=1750 lang=javascript
 *
 * [1750] Minimum Length of String After Deleting Similar Ends
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var minimumLength = function (s) {
    if (s.length < 2) {
        return s.length;
    }

    let left = 0;
    let right = s.length - 1;
    let leftLetter = s[left];

    while (true) {
        if (s[left] !== s[right]) {
            return right - left + 1;
        }
        if (left === right) {
            return 1;
        }

        while (s[left] === leftLetter) {
            left++;
            // s[left:right+1] 区间全都是相同的字母，全部被消除
            if (left === right + 1) {
                return 0;
            }
        }

        while (s[right] === leftLetter) {
            right--;
        }

        leftLetter = s[left];
    }
};
// @lc code=end

File: Daily/1750.minimum-length-of-string-after-deleting-similar-ends.cpp
/*
 * @lc app=leetcode id=1750 lang=cpp
 *
 * [1750] Minimum Length of String After Deleting Similar Ends
 */

#include <string>
using std::string;

// @lc code=start
class Solution {
 public:
  int minimumLength(string s) {
    string::size_type left = 0;
    string::size_type right = s.size() - 1;

    while (left < right && s[left] == s[right]) {
      char currentChar = s[left];
      while (s[left] == currentChar && left <= right) {
        left++;
      }

      while (s[right] == currentChar && left <= right) {
        right--;
      }
    }

    return right - left + 1;
  }
};
// @lc code=end

int main() {
  Solution sol;
  sol.minimumLength("ccbcbcbcc");
}